File: graph_search/topological_sort/210_course_schedule_II.cc
#include <algorithm>
#include <queue>
#include <vector>

enum class NodeStatus {
  kUnvisited,
  kSearching,
  kSearched,
};

// 存储有向图
std::vector<std::vector<int>> edges;
// 标记每个节点的状态
std::vector<NodeStatus> node_status;
// 用数组模拟栈
std::vector<int> result;
// 判断有向图中是否有环
bool valid = true;

void DFS(int u) {
  node_status[u] = NodeStatus::kSearching;
  // 搜索相邻节点, 如果有环, 则立刻停止搜索.
  for (int v : edges[u]) {
    if (node_status[v] == NodeStatus::kUnvisited) {
      // 如果未搜索, 那么搜索相邻节点.
      DFS(v);
      if (!valid) {
        return;
      }
    } else if (node_status[v] == NodeStatus::kSearching) {
      // 如果搜索中, 说明找到了环.
      valid = false;
      return;
    }
  }
  // 节点标记成已完成并入栈.
  node_status[u] = NodeStatus::kSearched;
  result.push_back(u);
}

// DFS: 用一个栈来存储所有已经搜索完成的节点
// 关键存储节点的访问状态和判断是否有环.
std::vector<int> FindOrder(int num_course,
                           const std::vector<std::vector<int>> prerequisites) {
  edges.resize(num_course);
  node_status.resize(num_course);
  for (const auto& prerequisite : prerequisites) {
    edges[prerequisite[1]].push_back(prerequisite[0]);
  }
  for (int i = 0; i < num_course; ++i) {
    if (node_status[i] == NodeStatus::kUnvisited) {
      DFS(i);
    }
    if (!valid) {
      return {};
    }
  }
  std::reverse(result.begin(), result.end());
  return result;
}

std::vector<int> indeg;
// BFS: 关键是想到利用入度来辅助判断该节点(课程)是否可以选择了.
std::vector<int> FindOrder(int num_course,
                           const std::vector<std::vector<int>> prerequisites) {
  edges.resize(num_course);
  indeg.resize(num_course);
  for (const auto& prerequisite : prerequisites) {
    edges[prerequisite[1]].push_back(prerequisite[0]);
    ++indeg[prerequisite[0]];
  }

  std::queue<int> q;
  for (int i = 0; i < num_course; ++i) {
    // 将所有入度为0的节点放入队列中
    if (indeg[i] == 0) {
      q.push(i);
    }
  }

  while (!q.empty()) {
    int u = q.front();
    q.pop();
    result.push_back(u);
    for (int v : edges[u]) {
      --indeg[v];
      if (indeg[v] == 0) {
        // 此时可以选v对应的课程了.
        q.push(v);
      }
    }
  }

  if (result.size() != num_course) {
    return {};
  }
  return result;
}

File: graph/topological_sort/210_course_schedule_II.py
from collections import defaultdict, deque
from typing import List


# 2026/05/21: 直接抄答案
# 深度优先遍历, DFS.
def find_order1(num_courses: int, prerequisites: List[List[int]]) -> List[int]:
    edges = defaultdict(list)  # 注意这里 edges 的初始化方式.
    visited = [0] * num_courses  # 0 = 未搜索, 1 = 搜索中, 2 = 已完成.
    result = list()  # 用数组模拟栈
    valid = True  # 判断是否有环

    for prerequisite in prerequisites:
        edges[prerequisite[1]].append(prerequisite[0])

    def dfs(u: int) -> None:
        nonlocal valid
        visited[u] = 1
        for v in edges[u]:
            if visited[v] == 0:
                dfs(v)
                if not valid:
                    return
            elif visited[v] == 1:
                valid = False
                return
        visited[u] = 2
        result.append(u)

    for i in range(num_courses):
        if valid and not visited[i]:
            dfs(i)

    if not valid:
        return list()

    return result[::-1]  # 下标 0 为栈底, 所以需要反序输出.


# 2026/05/21: 直接抄答案
# 广度优先遍历, BFS.
def find_order2(num_courses: int, prerequisites: List[List[int]]) -> List[int]:
    edges = defaultdict(list)
    indeg = [0] * num_courses  # 关键: 存储每个节点的入度.
    result = []

    for prerequisite in prerequisites:
        edges[prerequisite[1]].append(prerequisite[0])
        indeg[prerequisite[0]] += 1

    queue = deque([u for u in range(num_courses) if indeg[u] == 0])

    while queue:  # 判空可以直接写, 不需要判断 len 长度.
        u = queue.popleft()
        result.append(u)
        for v in edges[u]:
            indeg[v] -= 1
            if indeg[v] == 0:
                queue.append(v)

    if len(result) != num_courses:
        return []

    return result


num_courses = 4
prerequisites = [[1, 0], [2, 0], [3, 1], [3, 2]]
print(find_order1(num_courses, prerequisites))
